sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2322,7 +2322,7 @@ class Dataset[T] private[sql](
     }
     val attrs = this.logicalPlan.output
     val colsAfterDrop = attrs.filter { attr =>
-      attr != expression
+      !attr.semanticEquals(expression)
```

Comment:
Is there no palce having the same issue?

Reply:
Went through Dataset.scala - didn't find similar issue. However there might be the same problems in other places in our SQL code..

Reply:
There are other places. Please see https://github.com/apache/spark/pull/21449.

Reply:
ok, thanks. I think its ok to only target dataset.drop in this pr.

Reply:
+1 for @maropu 's opinion.